### PR TITLE
Authentication and Authorization

### PR DESCRIPTION
**Pull request by samuelgrant**
_This pull request is a work in progress._

---

**Features:**

This feature includes a set of views, controllers and models.

- Authentication
   - Login - Email/Password login, by default a login cookie lasts 30 minutes but this can be changed in appsettings.json
   - Request Password Reset - Allows a user to enter their email address to request a password reset
   - Reset Password - Change a user's password
   - Registration -> Authenticated users can create a new user's account, a random password will be generated, this locks the user out until they reset their own password. 

- User/Account Pages
  - Account page allows a user to change their details, and change their password
     - Not Yet Implemented; mailing lists that a user can subscribe to
  - Users page allows a user to edit other details about a user, register a new account, flag an account for password reset or delete an account

**To Do:**
- [x] Send a new account email with a 24-hour password reset link to a new user so they can gain access to their account
- [x] Create an account settings page so users can update their own name, email, phone number & password
  - [x] Update the accounts table to have a phone number string field
- [x] Create Users manage users page

### CODE REVIEW

File: appsettings.json [hunk 1]
```diff
@@ -1,6 +1,9 @@
 {
   "ConnectionStrings": {
-    "MsSqlConnection": "Server=(localdb)\\MSSQLLocalDB;Database=DeepcoveWebsite",
+    "MsSqlConnection": "Server=(localdb)\\MSSQLLocalDB;Database=DeepcoveWebsite;"
+  },
+  "LoginSettings": {
+    "PasswordRestTokenLength":  30
```
Comment (jacksos101) [inline appsettings.json]: typo - matches earlier typo

File: Controllers/Authentication/RegistrationController.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            if(account != null)
+                return BadRequest("An account with that email address already exists");
+
+            try
+            {
+                account = new Account
+                {
+                    Email = request.Str("email"),
+                    Name = request.Str("name"),
+                    Active = true,
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Controllers/Authentication/RegistrationController.cs]: Resolved

File: Properties/launchSettings.json [hunk 3]
```diff
@@ -21,7 +21,7 @@
       "environmentVariables": {
         "ASPNETCORE_ENVIRONMENT": "Development"
       },
-      "applicationUrl": "https://localhost:5001;http://localhost:5000"
+        "applicationUrl": "https://*:5001;http://*:5000"
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Properties/launchSettings.json]: Yes and it does work for me. We can take a look later

File: Controllers/Authentication/RequestPasswordResetController.cs [hunk 4]
```diff
@@ -2,17 +2,67 @@
...
+            try
+            {
+                Account account = await _Db.Accounts.Where(c => c.Email == request.Str("email")).FirstOrDefaultAsync();
+                if (account != null)
+                {
+                    PasswordReset reset = new PasswordReset
+                    {
+                        Account = account,
+                        Token = Utils.RandomString(20),
+                        ExpiresAt = DateTime.UtcNow.AddMinutes(_Configuration.GetValue<int>("PasswordRestTokenLength"))
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Controllers/Authentication/RequestPasswordResetController.cs]: Added an `e`

File: appsettings.json (hunk 1 above)
Comment (samuelgrant) [reply appsettings.json]: Resolved

File: Properties/launchSettings.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Properties/launchSettings.json]: Let's see if this is fixed by Windows 10... The issue was that we were binding to `localhost` which doesn't work with docker, at least in this instance. So instead I'm doing what we would do on a linux deployment and bind to any URL at a specific port. This should work for you, if it continues to error we can take a look.

File: Controllers/Authentication/RequestPasswordResetController.cs [hunk 5]
```diff
@@ -2,17 +2,67 @@
...
         }
+
+        public async Task<IActionResult> Index(IFormCollection request)
+        {
+            try
+            {
+                Account account = await _Db.Accounts.Where(c => c.Email == request.Str("email")).FirstOrDefaultAsync();
+                if (account != null)
+                {
+                    PasswordReset reset = new PasswordReset
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Controllers/Authentication/RequestPasswordResetController.cs]: Done with this code;
```CS
foreach (PasswordReset resetToken in await _Db.PasswordReset.Where(w => w.Account.Id == User.AccountId()).ToListAsync())
     resetToken.ExpiresAt = DateTime.UtcNow;
```

File: Controllers/Authentication/ResetPasswordController.cs [hunk 6]
```diff
@@ -2,17 +2,71 @@
...
+        public ResetPasswordController(Data.WebsiteDataContext db)
+        {
+            _Db = db;
+            _Hasher = new PasswordHasher<Account>();
+        }
+
         [HttpGet]
         public IActionResult Index(string token)
         {
+            // Redirect authenticated users
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Controllers/Authentication/ResetPasswordController.cs]: I wasn't going to do that, I'm not overly used to using password resets, I was only going to check it once when the user hits the submit button.

File: Controllers/Authentication/LoginController.cs [hunk 7]
```diff
@@ -1,26 +1,84 @@
...
             return View(viewName: "~/Views/Authentication/Login.cshtml");
         }
 
+        [HttpPost]
+        public async Task<IActionResult> Index(IFormCollection request)
+        {
+            Account account = await _Db.Accounts.Where(c => c.Email ==  request.Str("email")).FirstOrDefaultAsync();
+
+            //IF Username or password is invalid;
+            if (account == null || account.Password == _Hasher.HashPassword(account, request.Str("password")))
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Controllers/Authentication/LoginController.cs]: I've worked out how this works (I think), the issue should be resolved.

File: Properties/launchSettings.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (samuelgrant) [reply Properties/launchSettings.json]: @jacksos101 Is this issue resolved?

File: Resources/js/React/Components/FormControl.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    componentWillReceiveProps(nextProps) {
+        if (this.state.value != nextProps.value)
+            this.setState({ value: nextProps.value });
+    }
+  
+    
+    render() {
+        let id = false;
+        if (this.props.id)
+            id = this.props.id;
```
Comment (jacksos101) [inline Resources/js/React/Components/FormControl.js]: Would something like this work?

`let id = this.props.id || false; `

**Comment by samuelgrant** [comment]
@jacksos101 @MitchellQ - Let's get one last full code review over everything plus a practicle test from one of you.

Once I get this I'm happy for it to be merged

**Pull request merged**